### PR TITLE
Soft Exponential activation function

### PR DESCRIPTION
**Pull request by hobson**
softexp_activation

### CODE REVIEW

**Comment by hobson** [comment]
Issue #3842

**Comment by hobson** [comment]
Inviting a keras pro ( @uglyboxer) to help fix my broken `K.select`

File: keras/objectives.py [hunk 1]
```diff
@@ -1,6 +1,12 @@
 from __future__ import absolute_import
 import numpy as np
 from . import backend as K
+from .utils.generic_utils import get_from_module
+
+
+def constant_zero(y_true, y_pred, constant_loss=0.):
```
Comment (fchollet) [inline keras/objectives.py]: Let's not include that.

File: keras/layers/advanced_activations.py [hunk 2]
```diff
@@ -167,6 +173,77 @@ def get_config(self):
...
+        super(ParametricSoftExp, self).__init__(**kwargs)
+
+    def build(self, input_shape):
+        input_shape = input_shape[1:]
+        try:
+            self.alphas = K.variable(self.alpha_init * np.ones(input_shape),
+                                     name='{}_alphas'.format(self.name))
+        except:
+            self.alphas = K.variable(self.alpha_init, name='{}_alphas'.format(self.name))
+        assert self.alphas.shape == input_shape, "alphas.shape = {} != {} = input_shape".format(self.alphas.shape, input_shape)
```
Comment (fchollet) [inline keras/layers/advanced_activations.py]: Line too long. Prefer raising a `ValueError`. Use a consistent quote char (`'`).

**Comment by hobson** [comment]
definitely not. my mistake.

--Hobson
(503) 974-6274
gh <https://github.com/hobson/> twtr <https://twitter.com/hobsonlane> li
<https://www.linkedin.com/in/hobsonlane> g+
<http://plus.google.com/+HobsonLane/> so
<http://stackoverflow.com/users/623735/hobs>

File: keras/layers/advanced_activations.py [hunk 3]
```diff
@@ -167,6 +173,77 @@ def get_config(self):
...
+
+    def call_alpha_pos(self, x, alpha):
+        return alpha + (K.exp(alpha * x) - 1.) / alpha
+
+    def call_alpha_neg(self, x, alpha):
+        return - K.log(1 - alpha * (x + alpha)) / alpha
+
+    def call(self, x, mask=None):
+        pos = self.call_alpha_pos(x, self.alphas)
+        neg = self.call_alpha_neg(x, self.alphas)
```
Comment (fchollet) [inline keras/layers/advanced_activations.py]: Put these inside the `call` method, they shouldn't be class methods.

File: keras/objectives.py (hunk 1 above)
Comment (hobson) [inline keras/objectives.py]: Deleted

File: keras/layers/advanced_activations.py [hunk 4]
```diff
@@ -167,6 +173,77 @@ def get_config(self):
         return dict(list(base_config.items()) + list(config.items()))
 
 
+class ParametricSoftExp(Layer):
```
Comment (hobson) [inline keras/layers/advanced_activations.py]: Done

File: keras/layers/advanced_activations.py (hunk 2 above)
Comment (hobson) [inline keras/layers/advanced_activations.py]: eliminated this option... until the basic version works

File: keras/layers/advanced_activations.py (hunk 3 above)
Comment (hobson) [inline keras/layers/advanced_activations.py]: done

File: keras/backend/tensorflow_backend.py [hunk 5]
```diff
@@ -1291,6 +1293,19 @@ def switch(condition, then_expression, else_expression):
...
+    """TensorFlow API: tf.select(condition, t, e, name=None)
+
+    `condition`, `then_expression` & `else_expression`
+    must all 3 be tensors of the same shape
+    """
+    # tf.expand_dims(condition, 0)
+    condition = tf.cast(tf.ones_like(then_expression) * tf.to_float(condition), dtype=bool)
+    condition.set_shape(copy.copy(then_expression.get_shape()))
+    # condition = tf.tile(condition, tf.pack([1, tf.shape(then_expression)[1]]))
+    return tf.select(condition, then_expression, else_expression)
```
Comment (hobson) [inline keras/backend/tensorflow_backend.py]: @uglyboxer: when I fit to a ParametricSoftExponential() layer/activation, this fills tensor output with NaNs

File: keras/layers/advanced_activations.py [hunk 6]
```diff
@@ -1,8 +1,13 @@
+# -*- coding: utf-8 -*-
 from .. import initializations
 from ..engine import Layer
 from .. import backend as K
 import numpy as np
 
+import logging
+
+logger = logging.getLogger(__name__)
```
Comment (fchollet) [inline keras/layers/advanced_activations.py]: You don't need these two. Debugging leftovers?

**Comment by keunwoochoi** [comment]
@hobson Is there no update? I was about to implement it then found out this PR.

**Comment by keunwoochoi** [comment]
How should this function be defined for `x <= 0` when `alpha == -1`? Or probably shouldn't be, so  `-1 < alpha <= 1`?

**Comment by hobson** [comment]
The [plots in the feature request](
https://github.com/fchollet/keras/issues/3842) are helpful. Perhaps -INF or
it should be clipped at some large magnitude. I tried clipping at [-1e12,
+1e12] for some runs of toynet.play() but couldn't find a value that
consistently converged the net to reasonable accuracy for the multiply
problem.

--Hobson
(503) 974-6274
gh <https://github.com/hobson/> twtr <https://twitter.com/hobsonlane> li
<https://www.linkedin.com/in/hobsonlane> g+
<http://plus.google.com/+HobsonLane/> so
<http://stackoverflow.com/users/623735/hobs>

**Comment by keunwoochoi** [comment]
Thanks. Do you mean, at the end, the result is not stable in Theano as well?
What I want to implement exactly is a trainable log function and your code definitely helps, thanks for the work.

**Comment by hobson** [comment]
Correct. A 2x2 network of this implementation of softexponentials with
Theano as backend does not converge reliably to the multiply operation,
clipped or unclipped.

**Comment by keunwoochoi** [comment]
I see. Hopefully `parametric_log` is more stable. Someone in reddit also reported about the unstability, so without any bug it might be still like that.

**Comment by hobson** [comment]
Look forward to seeing your test results.

**Comment by fchollet** [comment]
Closing outdated PR. If you still care about the content of the PR, please submit a new PR to `master`, updated for the Keras 2.0 API.
